app/src/org/commcare/android/javarosa/PollSensorController.java

```diff
@@ -112,27 +111,43 @@ public void onLocationChanged(Location location) {
         }
     }
 
-    @Override
-    public void onProviderDisabled(String provider) {
+    private void sendBroadcast(Context context) {
+        Intent noGPSIntent = new Intent(GeoUtils.ACTION_CHECK_GPS_ENABLED);
+        context.sendStickyBroadcast(noGPSIntent);
     }
 
     @Override
-    public void onProviderEnabled(String provider) {
+    public void missingPermissions() {
+        missingPermissions = true;
+        Context context = CommCareApplication.instance();
+        sendBroadcast(context);
     }
 
     @Override
-    public void onStatusChanged(String provider, int status, Bundle extras) {
+    public void onLocationRequestFailure(@NotNull CommCareLocationListener.Failure failure) {
+        Context context = CommCareApplication.instance();
+        if (failure instanceof CommCareLocationListener.Failure.ApiException) {
+            Exception exception = ((CommCareLocationListener.Failure.ApiException) failure).getException();
+            if (exception instanceof ResolvableApiException) {
+                apiException = (ResolvableApiException) exception;
+                sendBroadcast(context);
+            } else {
+                // ignore and return, we can't do anything.
+            }
+        } else {
+            noProviders = true;
+            context.registerReceiver(
```

Comment:
where does this receiver get fired from ?

Reply:
Yeah, so this else part comes when the providers weren't enabled for this device so it sends a broadcast to the activity to ask users to enable providers `GeoUtils.goToProperLocationSettingsScreen`. 
So we register a receiver to listen to events when the providers are changed by the users from settings.